lib/tilt/template.rb

```diff
@@ -267,13 +262,25 @@ def compiled_method(locals_keys, scope_class=nil)
     end
 
     def local_extraction(local_keys)
-      local_keys.map do |k|
+      # If there is a locals key itself named `locals`, delete it from the ordered keys so we can
+      # assign it last. This is important because the assignment of all other locals depends on the
+      # `locals` local variable still matching the `locals` method argument given to the method
+      # created in `#compile_template_method`.
+      locals_local_key = local_keys.delete(:locals)
```

Comment:
Finding a good name for a variable like this is, as you can imagine, kind of tricky. I'm not precious about what I've picked here, and would be happy for you to make specific suggestions or tweak things directly.